internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,334 @@
+	for _, endpt := range r.targetResolverState.Endpoints {
+		var addrs []resolver.Address
+		for _, proxyAddr := range r.proxyAddrs {
+			for _, targetAddr := range endpt.Addresses {
+				addrs = append(addrs, proxyattributes.Set(proxyAddr, proxyattributes.Options{
+					User:        user,
+					ConnectAddr: targetAddr.Addr,
+				}))
+			}
+		}
+		endpoints = append(endpoints, resolver.Endpoint{Addresses: addrs})
+	}
+	r.curState.Addresses = addresses
+	r.curState.Endpoints = endpoints
+	return r.cc.UpdateState(r.curState)
+}
+
+// updateProxyResolverState updates the proxy resolver state by storing proxy
+// addresses and endpoints, marking the resolver as ready, and triggering a
+// state update if both proxy and target resolvers are ready. If the ClientConn
+// returns a non-nil error, it calls `ResolveNow()` on the target resolver.  It
+// is a StateListener function of wrappingClientConn passed to the proxy resolver.
+func (r *delegatingResolver) updateProxyResolverState(state resolver.State) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if logger.V(2) {
+		logger.Infof("Addresses received from proxy resolver: %s", state.Addresses)
+	}
+	if len(state.Endpoints) > 0 {
+		r.proxyAddrs = nil
+		for _, endpoint := range state.Endpoints {
+			r.proxyAddrs = append(r.proxyAddrs, endpoint.Addresses...)
+		}
+	} else {
+		r.proxyAddrs = state.Addresses
+	}
+	r.proxyResolverReady = true
+	err := r.updateClientConnStateLocked()
+	// Another possible approach was to block until updates are received from
+	// both resolvers. But this is not used because calling `New()` triggers
+	// `Build()`  for the first resolver, which calls `UpdateState()`. And the
+	// second resolver hasn't sent an update yet, so it would cause `New()` to
+	// block indefinitely.
+	if err != nil {
+		r.targetResolver.ResolveNow(resolver.ResolveNowOptions{})
+	}
+	return err
+}
+
+// updateTargetResolverState updates the target resolver state by storing target
+// addresses, endpoints, and service config, marking the resolver as ready, and
+// triggering a state update if both resolvers are ready. If the ClientConn
+// returns a non-nil error, it calls `ResolveNow()` on the proxy resolver. It
+// is a StateListener function of wrappingClientConn passed to the target resolver.
+func (r *delegatingResolver) updateTargetResolverState(state resolver.State) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if logger.V(2) {
+		logger.Infof("Addresses received from target resolver: %v", state.Addresses)
+	}
+	r.targetResolverState = state
+	r.targetResolverReady = true
+	// Update curState to include other state information, such as the
+	// service config, provided by the target resolver. This ensures
+	// curState contains all necessary information when passed to
+	// UpdateState. The state update is only sent after both the target and
+	// proxy resolvers have sent their updates, and curState has been
+	// updated with the combined addresses.
+	r.curState = state
```

Comment:
I don't think we want a separate `curState` field in the resolver.

You can instead incorporate this part of the logic into `updateClientConnStateLocked()`.  It can start by copying out the `targetResolverState`, then update the addresses and endpoints based on that and the proxy's state.

We should try to eliminate fields in structs as much as possible.  Don't have them "as a convenience".  Every piece of state you store has a complexity cost.  It makes things harder to understand, and bugs are more likely.

Reply:
Done.